### PR TITLE
Issue 1186 - Add registrar privilege support to secure ECA.RegisterUser in member services

### LINKED ISSUE #1186: Member services register user is insecure
ECAA.RegisterUser should only be allowable by a user with special privileges.
For example, the 'hyperledger.registrar' permission in a tcert could be required.
This implies that there must be one or more users (i.e. registrars) pre-registered by including one or more enrollment entries similar to:
    {'id': 'myID', 'secret': 'xxxx', roles: ['hyperledger.registrar']}

### PR DESCRIPTION
**Pull request by smithbk**
## Description

1) Store additional metadata in user table of member services including the types of roles that a member can register (if any) and those that it can delegate to others to register.  See the additional comments in membersrvc.yaml file.
2) Enforce this policy in the ECAA.RegisterUser GRPC call, which includes checking signature and policy of the caller.
3) Update node client SDK to pass additional info in GRPC call, including signature, roles, and delegate roles.
## Motivation and Context

ECAA.RegisterUser is a security exposure today as there are no security checks made.
This closes this exposure.

Fixes #1186
## How Has This Been Tested?

I added a new test case in the node client SDK at fabric/sdk/node/test/unit/registrar.js and updated fabric/sdk/node/README.md with this info.

## Checklist:

- [x] I have added a [Signed-off-by](https://github.com/hyperledger/fabric/blob/master/CONTRIBUTING.md#legal-stuff)
- [x] Either no new documentation is required by this change, OR I added new documentation
- [x] Either no new tests are required by this change, OR I added new tests
- [x] I have run [goimports](https://godoc.org/golang.org/x/tools/cmd/goimports), [go vet](https://golang.org/cmd/vet/), and [golint](https://github.com/golang/lint). I have cleaned up all valid errors and warnings in code I have added or modified. These tools may generate false positives. Don't be worried about ignoring some errors or warnings. The goal is clean, consistent, and readable code.

Signed-off-by: Keith Smith bksmith@us.ibm.com

### CODE REVIEW

**Comment by binhn** [comment]
It looks like the build system is going mad

**Comment by rameshthoomu** [comment]
@binhn: Please ignore Jenkins build results for now as we are setting up jenkins and testing functionality. Regarding Travis build result, someone has to investigate why unit-tests are failing.

File: sdk/node/test/unit/registrar.js [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+   //
+   // Configure the test chain
+   //
+   chain.setKeyValStore(hlc.newFileKeyValStore('/tmp/keyValStore'));
+   chain.setMemberServicesUrl("grpc://localhost:50051");
+   chain.getMember("admin", function (err, admin) {
+      if (err) return cb(err);
+      // Enroll the admin
+      admin.enroll("Xurw3yU9zI0l", function (err, crypto) {
+          if (err) return fail(t,"Failed to enroll WebAppAdmin member",err);
```
Comment (adecaro) [inline sdk/node/test/unit/registrar.js]: t is undefined here. Shall registrarTest take it in input?

**Comment by adecaro** [comment]
@smithbk, @diegomasini To stress test the member services and the registrar, I have executed two consecutive times the following commands:

```
export DEBUG=hlc,crypto && clear && rm -rf /tmp/keyValStore/ && node test/unit/registrar.js
```

The result is the crash of the member services. 

Now, I don't think the crash depends on this PR, but anyway the member services should not crash. It should just report an error. 

This is the stack trace I see.

```
TRACE: 2016/06/05 10:09:46 eca.go:221: gRPC ECAP:CreateCertificate
TRACE: 2016/06/05 10:09:46 ca.go:628: Reading token for admin.
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x20 pc=0x4f585b]

goroutine 38 [running]:
panic(0xae0de0, 0xc82000a140)
    /usr/local/go/src/runtime/panic.go:481 +0x3e6
github.com/hyperledger/fabric/membersrvc/ca.(*ECAP).CreateCertificatePair(0xc820028c38, 0x7f74dd5c01e8, 0xc8203960c0, 0xc8203961e0, 0xc820034b58, 0x0, 0x0)
    /home/angelo/golang/src/github.com/hyperledger/fabric/membersrvc/ca/eca.go:232 +0x269b
github.com/hyperledger/fabric/membersrvc/protos._ECAP_CreateCertificatePair_Handler(0xb2f940, 0xc820028c38, 0x7f74dd5c01e8, 0xc8203960c0, 0xc8203961b0, 0x0, 0x0, 0x0, 0x0)
    /home/angelo/golang/src/github.com/hyperledger/fabric/membersrvc/protos/ca.pb.go:1112 +0x132
github.com/hyperledger/fabric/vendor/google.golang.org/grpc.(*Server).processUnaryRPC(0xc8202b0360, 0x7f74dd606ba8, 0xc8202e6d80, 0xc820329800, 0xc8202b5ac0, 0x1111398, 0xc820396090, 0x0, 0x0)
    /home/angelo/golang/src/github.com/hyperledger/fabric/vendor/google.golang.org/grpc/server.go:350 +0x847
github.com/hyperledger/fabric/vendor/google.golang.org/grpc.(*Server).handleStream(0xc8202b0360, 0x7f74dd606ba8, 0xc8202e6d80, 0xc820329800, 0xc820396090)
    /home/angelo/golang/src/github.com/hyperledger/fabric/vendor/google.golang.org/grpc/server.go:467 +0x743
github.com/hyperledger/fabric/vendor/google.golang.org/grpc.(*Server).Serve.func2.1.1(0xc8202b0360, 0x7f74dd606ba8, 0xc8202e6d80, 0xc820329800, 0xc820396090, 0xc820314e70)
    /home/angelo/golang/src/github.com/hyperledger/fabric/vendor/google.golang.org/grpc/server.go:278 +0x49
created by github.com/hyperledger/fabric/vendor/google.golang.org/grpc.(*Server).Serve.func2.1
    /home/angelo/golang/src/github.com/hyperledger/fabric/vendor/google.golang.org/grpc/server.go:280 +0x5ac
```

**Comment by smithbk** [comment]
@adecaro @diegomasini The last commit fixed both issues Angelo found.  On the 1st, I'm not sure why jshint did not find it.  On the 2nd, there was a bug in a trace statement which is fixed, but Angelo I assume you know that running this test again after removing /tmp/keyValStore will fail but that is correct because admin (and other users) have already been enrolled.  You need to also remove /var/hyperledger/production and restart membersrvc for it to work.  It is a valid test to rerun node /test/unit/registrar.js without removing /tmp/keyValStore.

**Comment by smithbk** [comment]
I'm not sure why go isn't in the path of jenkins/hudson.  Anyone?

[fabric-github-verify-z] $ /bin/bash -eu /tmp/hudson7007350690855464636.sh
which: no go in (/opt/go/go1.6.linux.amd64/bin:/w/workspace/fabric-github-verify-z/gopath/bin:/usr/local/bin:/usr/bin)
Makefile:43: **\* "No go in PATH: Check dependencies".  Stop.
Build step 'Execute shell' marked build as failure

**Comment by christo4ferris** [comment]
@smithbk etc please disregard the jenkins build errors... the LF team and a few others are still in the process of getting that fully established.

**Comment by adecaro** [comment]
@smithbk, I was actually intentionally running the commands in the wrong way to see if the fabric was able to not crash. IMHO, we need to stress the fabric more and more also in this way.

But I can confirm that the tests pass running the proper commands :)

**Comment by smithbk** [comment]
@adecaro I added a test case for this to registrar.js

**Comment by binhn** [comment]
@diegomasini Could you review.

**Comment by diegomasini** [comment]
@binhn @smithbk it looks good to me

**Comment by binhn** [comment]
@smithbk please rebase

**Pull request merged**